lib/gollum/public/gollum/javascript/editor/modes.js.erb

```diff
@@ -0,0 +1,151 @@
+/*  modes.js
+
+    Add highlighting rules for gollum-specific block level syntax to ACE.
+    We first have to extends the Highlighters for each supported syntax,
+    so that later, we can extend the relevant ACE Mode to use the custom Highlighter.
+    We do this for all and only those languages which ACE supports, i.e.,
+    for which it has a specific Highlighter and Mode. So we do not extend e.g. Creole,
+    for which there is no ACE Mode (we use the 'text mode' instead).
+    At the moment there are two gollum-wide block level syntaxes which need to be added:
+
+    * GitHub style code blocks: ``` ... ```
+    * UML blocks: @startuml ... @enduml
+
+    The regexes defined for these below should be equivalent to the regexes used by
+    gollum-lib to extract these blocks.
+
+    ACE Docs:
+      * Highlight Rules: https://cloud9-sdk.readme.io/docs/highlighting-rules
+      * Dynamically generating modes: https://stackoverflow.com/questions/22166784/dynamically-update-syntax-highlighting-mode-rules-for-the-ace-editor
+*/
+
+<%
+codeblock_rule_name = 'GitHubBlock'
+codeblock_rule_id = 'githubblock'
+%>
+
+// The rule defining the github-codeblock state for the ACE Tokenizer
+var <%= codeblock_rule_name %> = [{
+  token: "support.function",
+  regex: /^```$/,
+  next:  "start"
+}];
+
+// The rule defining how the ACE Tokenizer gets into the github-codeblock state
+var GitHubStart = {
+  token: "support.function",
+  regex: /^```/,
+  next:  "<%= codeblock_rule_id %>"
+};
+
+<%
+uml_rule_name = 'UMLBlock'
+uml_rule_id = 'umlblock'
+%>
+
+// The rule defining the UML state for the ACE Tokenizer
+var <%= uml_rule_name %>  = [{
+  token: "support.function",
+  regex: /^\@enduml/,
```

Comment:
Why only `@enduml` here?

Reply:
Once the tokenizer is in the `umlblock` state, we only need to look for the line that brings us back to the `start` state.

Actually, I just realized this means we're not marking the contents of UML/code blocks as code (although they will be prevented from being marked as different tokens). Maybe it would be nicer if the entire block, not just the start/stop syntax, was marked as code (and so also highlighted blue in the editor).

Reply:
> Maybe it would be nicer if the entire block, not just the start/stop syntax, was marked as code

Agreed.